File: go/2023/day01b.go
//go:build ignore

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"unicode"
)

func main() {
	fmt.Println(solve())
}

const words string = "one two three four five six seven eight nine"

func solve() int {
	scanner := bufio.NewScanner(bufio.NewReader(os.Stdin))

	var total int
	for scanner.Scan() {
		line := scanner.Text()

		var numbers []int
		for i := 0; i < len(line); i++ {
			if unicode.IsDigit(rune(line[i])) {
				num, _ := strconv.Atoi(string(line[i]))
				numbers = append(numbers, num)
			} else {
				for j, word := range strings.Split(words, " ") {
					if strings.HasPrefix(line[i:], word) {
						numbers = append(numbers, j+1)
						break
					}
				}
			}
		}

		total += numbers[0]*10 + numbers[len(numbers)-1]
	}

	return total
}

File: py/2023/day01b.py
import re

total = 0

words = "one two three four five six seven eight nine".split()


def convert_to_int(x: str) -> int:
    if x in words:
        return words.index(x) + 1
    else:
        return int(x)


for line in open(0):
    digits = re.findall(f"(?=({ '|'.join(words) }|\\d))", line)

    tens_place = convert_to_int(digits[0]) * 10
    ones_place = convert_to_int(digits[-1])

    total += tens_place + ones_place

print(total)
